src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidRunCommand.cs

```diff
@@ -0,0 +1,261 @@
+            logger.LogDebug($"Timeout = {_arguments.Timeout.TotalSeconds} seconds.");
+
+            // Package Name is not guaranteed to match file name, so it needs to be mandatory.
+            string apkPackageName = _arguments.PackageName;
+
+            var runner = new AdbRunner(logger);
+
+            // Make sure the adb server is started
+            runner.StartAdbServer();
+
+            // enumerate the devices attached and their architectures
+            // Tell ADB to only use that one (will always use the present one for systems w/ only 1 machine)
+            var deviceToUse = runner.GetDeviceToUse(logger, "package:" + apkPackageName, "app");
+
+            if (deviceToUse == null)
+            {
+                return Task.FromResult(ExitCode.ADB_DEVICE_ENUMERATION_FAILURE);
+            }
+
+            runner.SetActiveDevice(deviceToUse);
+
+            // Wait til at least device(s) are ready
+            runner.WaitForDevice();
+
+            // Empty log as we'll be uploading the full logcat for this execution
+            runner.ClearAdbLog();
+
+            logger.LogDebug($"Working with {runner.GetAdbVersion()}");
+
+            return Task.FromResult(InvokeHelper(
+                logger,
+                apkPackageName,
+                _arguments.InstrumentationName,
+                _arguments.InstrumentationArguments,
+                _arguments.OutputDirectory,
+                _arguments.DeviceOutputFolder,
+                _arguments.Timeout,
+                _arguments.ExpectedExitCode,
+                runner));
+        }
+
+        public ExitCode InvokeHelper(
+            ILogger logger,
+            string apkPackageName,
+            string? instrumentationName,
+            Dictionary<string, string> instrumentationArguments,
+            string outputDirectory,
+            string? deviceOutputFolder,
+            TimeSpan timeout,
+            int expectedExitCode,
+            AdbRunner runner)
+        {
+
+            int instrumentationExitCode = (int)ExitCode.GENERAL_FAILURE;
+
+            try
+            {
+                // No class name = default Instrumentation
+                ProcessExecutionResults? result = runner.RunApkInstrumentation(apkPackageName, instrumentationName, instrumentationArguments, timeout);
+                bool processCrashed = false;
+                bool failurePullingFiles = false;
+
+                using (logger.BeginScope("Post-test copy and cleanup"))
+                {
+                    if (result.ExitCode == (int)ExitCode.SUCCESS)
+                    {
+                        (var resultValues, var instrExitCode) = ParseInstrumentationOutputs(logger, result.StandardOutput);
+
+                        // This is where test instrumentation can communicate outwardly that test execution failed
+                        instrumentationExitCode = instrExitCode;
```

Comment:
Afaik we could have assigned the value in the previous line, but I might be mistaken.